crates/red_knot_python_semantic/src/types.rs

```diff
@@ -401,27 +402,35 @@ impl<'db> Type<'db> {
     ///     pass
     /// ```
     ///
-    /// Returns `None` if `self` represents a type that is not iterable.
-    fn iterate(&self, db: &'db dyn Db) -> Option<Type<'db>> {
-        // `self` represents the type of the iterable;
-        // `__iter__` and `__next__` are both looked up on the class of the iterable:
-        let type_of_class = self.to_meta_type(db);
-
-        let dunder_iter_method = type_of_class.member(db, "__iter__");
-        if !dunder_iter_method.is_unbound() {
-            let iterator_ty = dunder_iter_method.call(db)?;
-            let dunder_next_method = iterator_ty.to_meta_type(db).member(db, "__next__");
-            return dunder_next_method.call(db);
-        }
+    /// Emits a diagnostic and returns `Unknown` if `self` represents a type that is not iterable.
+    fn iterate(&self, db: &'db dyn Db) -> IterableElementTypeResult<'db> {
+        fn loop_var_from_iterable<'db>(
+            db: &'db dyn Db,
+            iterable_ty: &Type<'db>,
+        ) -> Option<Type<'db>> {
+            // `__iter__` and `__next__` are both looked up on the class of the iterable:
+            let iterable_meta_type = iterable_ty.to_meta_type(db);
+
+            let dunder_iter_method = iterable_meta_type.member(db, "__iter__");
+            if !dunder_iter_method.is_unbound() {
+                let iterator_ty = dunder_iter_method.call(db)?;
+                let dunder_next_method = iterator_ty.to_meta_type(db).member(db, "__next__");
+                return dunder_next_method.call(db);
+            }
 
-        // Although it's not considered great practice,
-        // classes that define `__getitem__` are also iterable,
-        // even if they do not define `__iter__`.
-        //
-        // TODO this is only valid if the `__getitem__` method is annotated as
-        // accepting `int` or `SupportsIndex`
-        let dunder_get_item_method = type_of_class.member(db, "__getitem__");
-        dunder_get_item_method.call(db)
+            // Although it's not considered great practice,
+            // classes that define `__getitem__` are also iterable,
+            // even if they do not define `__iter__`.
+            //
+            // TODO this is only valid if the `__getitem__` method is annotated as
+            // accepting `int` or `SupportsIndex`
+            let dunder_get_item_method = iterable_meta_type.member(db, "__getitem__");
+            dunder_get_item_method.call(db)
+        }
```

Comment:
We no longer need the inner function, now that there's no monomorphization happening in the function

Reply:
I kept the inner function for simplicity and readability, since there are several places where we return early, and it seemed easier with the question mark operator. But it probably doesn't add much any more, you're right.